# Ejercicios en clase de pytorch

In [8]:
pip uninstall torch torchvision torchaudio -y

Found existing installation: torch 2.5.1+cu121
Uninstalling torch-2.5.1+cu121:
  Successfully uninstalled torch-2.5.1+cu121
Found existing installation: torchvision 0.20.1+cu121
Uninstalling torchvision-0.20.1+cu121:
  Successfully uninstalled torchvision-0.20.1+cu121
Found existing installation: torchaudio 2.5.1+cu121
Uninstalling torchaudio-2.5.1+cu121:
  Successfully uninstalled torchaudio-2.5.1+cu121
Note: you may need to restart the kernel to use updated packages.


You can safely remove it manually.


In [1]:
import torch
import torch.nn as nn

autograd es calcular la distribucion de pesos por neuronas

optimizadores es el gradient descent para actualizar los pesos


In [2]:
import torch 
import numpy as np
x= torch.tensor([1,2,3])

array = np.array([1,2,3])
print(array) 
print(x)

x1 = torch.tensor(array)
print(x1)

[1 2 3]
tensor([1, 2, 3])
tensor([1, 2, 3])


In [3]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cuda


In [4]:
import torch

print(torch.cuda.is_available())
print(torch.cuda.device_count())
print(torch.version.cuda)

True
1
12.1


Creo que en este kernel de Python no tengo cuda, instalar

In [5]:
import torch
print(torch.version.cuda)

12.1


In [6]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2025 NVIDIA Corporation
Built on Wed_Aug_20_13:58:20_Pacific_Daylight_Time_2025
Cuda compilation tools, release 13.0, V13.0.88
Build cuda_13.0.r13.0/compiler.36424714_0


In [7]:
!nvidia-smi

Thu Mar 12 18:26:58 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 591.74                 Driver Version: 591.74         CUDA Version: 13.1     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 5060 ...  WDDM  |   00000000:02:00.0 Off |                  N/A |
| N/A   36C    P4             11W /   50W |       0MiB /   8151MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

El problema era que instale pytorch sin cuda, ya instale el pytorch con cuda 12.1 y ahora si agarro mi device

In [9]:
import torch.nn as nn

class FeedForwardNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super().__init__()
        self.hidden = nn.Linear(input_size, hidden_size)
        self.output = nn.Linear(hidden_size, output_size)
        self.ReLU = nn.ReLU()
        

    def forward(self,x):
        x = self.hidden(x)
        x = self.ReLU(x)
        x = self.output(x)
        return x
         

In [ ]:
momdel = nn.Sequential(
    nn.Linear(input_size, hidden_size),
    nn.ReLU

)

In [16]:
# Inicializar las funciones de activacion es sencillo:

nn.ReLU()     # Capas ocultas  
nn.Sigmoid()  # Clasificacion binaria
nn.Softmax()  # Clasificacion multiclase
nn.Tanh()     # Capas ocultas  

Tanh()

In [18]:
# 3 Funciones de perdida para PyTorch
nn.MSELoss()          # Regresion
nn.BCELoss()          # Clasificacion Binaria
nn.CrossEntropyLoss() # Clasificacion multiclase
nn.Sigmoid()

# CrossEntropyLoss no necesita Softmax, porque ya lo hace automaticamente la funcion de perdida
# BCELoss (Problemas de clasificacion binaria) requiere nn.Sigmoid() . el resultado solo me va a dar una probabilidad
# de si es positivo y ya, solo un escalar (de 0 a 1). Softmax si te da un vector

Sigmoid()

Caso de uso para el SoftMax() el usuario final quiere saber la probabilidad de la clasificacion

Pregunta para clase: # Puede una clasificacion binaria tener 2 neuroans en vez de solo 1?


**Funcion de perdida**

In [ ]:
# For regression
loss_fn = nn.MSELoss()
loss = loss_fn(pred, real)

loss.backward()   #hace el backpropragation para tener el erro de manera proporcional por neuronas


**Funcion de optimizacion**

In [ ]:
import torch.optim as optim

# Stochastic gradient descent
optimizer = optim.SGD(model.parameters(), lr=0.001) # Se le pasa el learning rate
# Adam
optimizer = optim.Adam(model.parameters(), lr=0.001)

# ESTO SE DEFINE ANTES DEL ENTRENAMIENTO

In [ ]:
# COMO SE EJECUTA DURANTE EL ENTRENAMIENTO?

# ANTES de iniciar el minibatch se usa zero grad (PONE EN 0 LAS GRADIENTES) [PARA QUE SOLO SE USEN LOS GRADIENTES DE ESTA ITERACION]
# La variable es global no especifica a un batch, por eso se reinicia despues de un batch
optimizer.zero_grad()

# DESPUES
loss.backward()  # HACEMOS el backpropragation del error proporcional a las neuronas 

# Ahora si se ejecuta el actualizar los parametros usando el optimizador en base al error.
optimizer.step()

**Dataloaders**

In [ ]:
from torch.utils.data import DataLoader, TensorDataset

X_tensor = torch.Tensor(X_train, dtype=torch.float32)
y_tensor = torch.Tensor(y_train, dtype=torch.float32)
dataset = TensorDataset(X_tensor, y_tensor)

dataloader= DataLoader(dataset, batch_size=32, shuffle= True)

for x_batch, y_batch in dataloader:
    # X_batch

**Entrenamiento**


In [ ]:
model= FeedForwardNN(input_size=10, hidden_size=32, output_size=1)
loss_fn = nn.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=0.001)

model.train()

for epoch in range(num_epochs):
    for x_batch, y_batch in dataloader:
        train_batch(model,x_)

In [ ]:
def train_batch(model, X_batch, y_batch, optimizer, loss_fn):
    # Forward pass
    predictions = model(X_batch)
    

**Evaluacion**

In [ ]:
model.eval()
with torch.no_grad():
    predictions = model(X_test)
    test_loss = loss_fn(predictions, y_test)

**Guardar y cargar modelos**

In [ ]:
# Se usa el state_dict() para poder guardar los pesos y sesgos de una red neuronal

torch.save(model.state_dict(), "Model.pth")

# para cargarlo es inicializar el model con los mismos estrcutura 
